Backend Attention Function

### Feature description

Add a multihead qkv attention function to burn and efficiently implement it for each backend.

### Feature motivation

Torch 2.0 uses flash attention by default which is much faster and uses less memory than high level implementations such as the one currently used in burn. Having burn call flash attention from tch would significantly improve performance on transformer-based models.

I tried following the example from the book but I get about 30 errors like:
```
error[E0277]: the trait bound `<B as ADBackend>::InnerBackend: backend::Backend` is not satisfied
   --> src/model/stablediffusion/mod.rs:191:10
    |
191 | #[derive(Module, Debug)]
    |          ^^^^^^ the trait `backend::Backend` is not implemented for `<B as ADBackend>::InnerBackend`
    |
note: required by a bound in `stablediffusion::LatentDecoder`
   --> src/model/stablediffusion/mod.rs:192:29
    |
192 | pub struct LatentDecoder<B: Backend> {
    |                             ^^^^^^^ required by this bound in `LatentDecoder`
    = note: this error originates in the derive macro `Module` (in Nightly builds, run with -Z macro-backtrace for more info)
```

The backend code is:
```
pub trait Backend: burn::tensor::backend::Backend {
    fn qkv_attention(q: Self::TensorPrimitive<3>,
        k: Self::TensorPrimitive<3>,
        v: Self::TensorPrimitive<3>,
        mask: Option<Self::TensorPrimitive<2>>,
        n_head: usize,) -> Self::TensorPrimitive<3> {
            qkv_attention(Tensor::<Self, 3>::from_primitive(q), Tensor::from_primitive(k), Tensor::from_primitive(v), mask.map(|m| Tensor::from_primitive(m)), n_head).into_primitive()
        }

    fn attn_decoder_mask(seq_length: usize, device: &Self::Device) -> Self::TensorPrimitive<2> {
        attn_decoder_mask::<Self>(seq_length, device).into_primitive()
    }
}

use burn_autodiff;

impl<B: Backend> Backend for burn_autodiff::ADBackendDecorator<B> {}
```

--- a/burn-derive/src/module/base.rs
+++ b/burn-derive/src/module/base.rs
@@ -116,21 +116,17 @@ fn fetch_backend_trait(generics: &syn::Generics) -> proc_macro2::TokenStream {
     - Any backend trait is supported.";
 
     for param in generics.params.iter() {
-        match &param {
-            syn::GenericParam::Type(ty) => {
-                if ty.ident == "B" {
-                    let bound = ty
-                        .bounds
-                        .first()
-                        .expect(BACKEND_TRAIT_COMPILATION_ERROR_MSG);
+        if let syn::GenericParam::Type(ty) = &param {
+            if ty.ident == "B" {
+                let bound = ty
+                    .bounds
+                    .first()
+                    .expect(BACKEND_TRAIT_COMPILATION_ERROR_MSG);
 
-                    return quote! {
-                        #bound
-                    }
-                    .into();
-                }
+                return quote! {
+                    #bound
+                };
             }
-            _ => {}
         }
     }
 

--- a/burn-derive/src/module/base.rs
+++ b/burn-derive/src/module/base.rs
@@ -21,6 +21,7 @@ pub(crate) fn derive_impl(ast: &syn::DeriveInput) -> TokenStream {
     }
 
     let (generics, generics_ty, generics_where) = ast.generics.split_for_impl();
+    let backend_trait = fetch_backend_trait(&ast.generics);
 
     let display_fn = display::display_fn(name);
 
@@ -51,7 +52,11 @@ pub(crate) fn derive_impl(ast: &syn::DeriveInput) -> TokenStream {
             #map_mut
         }
 
-        impl #generics burn::module::ADModule<B> for #name #generics_ty where B: burn::tensor::backend::ADBackend, {
+        impl #generics burn::module::ADModule<B> for #name #generics_ty
+        where
+            B: burn::tensor::backend::ADBackend,
+            <B as burn::tensor::backend::ADBackend>::InnerBackend: #backend_trait,
+        {
             type InnerModule=#name<B::InnerBackend, #generics_names_except_backend>;
 
             #valid_fn
@@ -102,6 +107,34 @@ fn constant_impl(ast: &syn::DeriveInput) -> TokenStream {
     };
 
     gen.into()
+}
+
+fn fetch_backend_trait(generics: &syn::Generics) -> proc_macro2::TokenStream {
+    static BACKEND_TRAIT_COMPILATION_ERROR_MSG: &str = "Modules should be generic over a backend.
+    - The generic argument named `B` should have its first trait bound being a backend trait.
+    - The default backend trait is `burn::tensor::backend::Backend`.
+    - Any backend trait is supported.";
+
+    for param in generics.params.iter() {
+        match &param {
+            syn::GenericParam::Type(ty) => {
+                if ty.ident == "B" {
+                    let bound = ty
+                        .bounds
+                        .first()
+                        .expect(BACKEND_TRAIT_COMPILATION_ERROR_MSG);
+
+                    return quote! {
+                        #bound
+                    }
+                    .into();
+                }
+            }
+            _ => {}
+        }
+    }
+
+    panic!("{BACKEND_TRAIT_COMPILATION_ERROR_MSG}");
 }
 
 fn generics_names_except_backend(generics: &syn::Generics) -> proc_macro2::TokenStream {

